### PR TITLE
MSC4108 implementation

### PR DESCRIPTION
**Pull request by sandhose**
This is a native implementation of MSC4108

It works by keeping the pushed content in an in-memory map. It means this is meant to be handled by the same worker.
This map has a fixed capacity, and entries are regularly evicted.

The capacity and various other knobs are hard-coded for now, and could be exposed in the config.
It also doesn't have any rate limiting, meaning you could force the expiration of other user's entries by spamming the creation of new request.
Both of those shortcomings could be handled in a separate pull request.

### CODE REVIEW

File: synapse/http/server.py [hunk 1]
```diff
@@ -909,7 +909,18 @@ def set_cors_headers(request: "SynapseRequest") -> None:
     request.setHeader(
         b"Access-Control-Allow-Methods", b"GET, HEAD, POST, PUT, DELETE, OPTIONS"
     )
-    if request.experimental_cors_msc3886:
+    if request.path is not None and request.path.startswith(
+        b"/_matrix/client/unstable/org.matrix.msc4108/rendezvous"
```
Comment (hughns) [inline synapse/http/server.py:913]: This path needs to match https://github.com/element-hq/synapse/pull/17056/files#diff-60b44b832bdef37a1bd6f9174c4e59ad6539a88363ad7292c4a48e9e556da4a7R81

Maybe it should be a shared constant?

**Review by hughns — CHANGES_REQUESTED** [review]

File: synapse/http/server.py (hunk 1 above)
Comment (hughns) [inline synapse/http/server.py:913]:
```suggestion
        b"/_synapse_/rendezvous"
```

Comment (sandhose) [reply synapse/http/server.py]: done!

Comment (sandhose) [reply synapse/http/server.py]: I fixed the path but it isn't using a constant

File: synapse/rest/client/rendezvous.py [hunk 2]
```diff
@@ -97,9 +97,25 @@ async def on_POST(self, request: SynapseRequest) -> None:
...
+    PATTERNS = client_patterns(
+        "/org.matrix.msc4108/rendezvous$", releases=[], v1=False, unstable=True
+    )
+
+    def __init__(self, hs: "HomeServer") -> None:
+        super().__init__()
+        self._handler = hs.get_rendezvous_handler()
+
+    def on_POST(self, request: SynapseRequest) -> None:
+        self._handler.handle_post(request)
```
[the remark this answers is not in the document]
Comment (hughns) [reply synapse/rest/client/rendezvous.py:110]: Ah, I forgot that you had put a note to do this in another PR.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Just some small comments, nothing major. Thanks for adding tests, and love all the rust here!

File: rust/src/rendezvous/mod.rs [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+            .ok_or_else(NotFoundError::new)?;
+
+        if !if_match.precondition_passes(&session.etag()) {
+            let mut headers = HeaderMap::new();
+            prepare_headers(&mut headers, session);
+
+            return Err(SynapseError::new(
+                StatusCode::PRECONDITION_FAILED,
+                "ETag does not match".to_owned(),
+                "M_CONCURRENT_WRITE",
```
[the remark this answers is not in the document]
Comment (sandhose) [reply rust/src/rendezvous/mod.rs:248]: That error code is defined by the MSC though, should it be prefixed or something?

https://github.com/matrix-org/matrix-spec-proposals/blob/element-hq/oidc-qr-login/proposals/4108-oidc-qr-login.md#send-a-payload-to-the-rendezvous-session-put-rendezvous-session-url

File: rust/src/rendezvous/mod.rs [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        }
+
+        // Generate a new ULID for the session from the current time.
+        let id = Ulid::from_datetime(now);
+
+        let uri = format!("{base}/{id}", base = self.base);
+
+        let body = request.into_body();
+
+        let session = Session::new(body, content_type, now, Duration::from_secs(5 * 60));
```
[the remark this answers is not in the document]
Comment (sandhose) [reply rust/src/rendezvous/mod.rs:168]: @hughns this means the QR code would expire after 30s and a new one should be generated? Or is it expected that GETing on the endpoint keeps it alive until the user scans it?

File: rust/src/rendezvous/mod.rs [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+            return Err(SynapseError::new(
+                StatusCode::PRECONDITION_FAILED,
+                "ETag does not match".to_owned(),
+                "M_CONCURRENT_WRITE",
+                None,
+                Some(headers),
+            ));
+        }
+
+        session.update(data, content_type, now);
```
[the remark this answers is not in the document]
Comment (hughns) [reply rust/src/rendezvous/mod.rs:254]: The MSC should say that the `PUT` doesn't bump the expiry. I will check that this is the case. But this implementation looks correct in this regard.

[the remark this answers is not in the document]
Comment (hughns) [reply rust/src/rendezvous/mod.rs:254]: I've corrected the MSC in https://github.com/matrix-org/matrix-spec-proposals/commit/4f9a4a42e6ba442989bee06a037eaedecc1cc846.

File: rust/src/rendezvous/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hughns) [reply rust/src/rendezvous/mod.rs:248]: In https://github.com/element-hq/synapse/pull/17113 I have changed `errcode` to be `M_UNKNOWN` and added a `org.matrix.msc4108.errcode` value of `M_CONCURRENT_WRITE`.

[the remark this answers is not in the document]
Comment (anoadragon453) [reply rust/src/rendezvous/mod.rs:248]: I'd like to see a reference to `org.matrix.msc4108.errcode` in the MSC, fwiw.

[the remark this answers is not in the document]
Comment (hughns) [reply rust/src/rendezvous/mod.rs:248]: > Sounds good, though I'd like to see a reference to `org.matrix.msc4108.errcode` in the MSC. Under ["Unstable Prefix"](https://github.com/matrix-org/matrix-spec-proposals/blob/element-hq/oidc-qr-login/proposals/4108-oidc-qr-login.md?rgh-link-date=2024-04-22T15%3A14%3A38Z#unstable-prefix).

Added in https://github.com/matrix-org/matrix-spec-proposals/commit/76f175b0110dbda4deb33c33c692a1a7202a42ba

**Review by anoadragon453 — APPROVED** [review]
This LGTM after https://github.com/element-hq/synapse/pull/17122, thanks!

Issues to track follow-up work:

- https://github.com/element-hq/synapse/issues/17124
- https://github.com/element-hq/synapse/issues/17123

**Comment by anoadragon453** [comment]
Tests were failing due to https://github.com/element-hq/synapse/pull/17125. This PR has now been updated with latest `develop`, and now all CI checks will run.

**Pull request merged**